### PR TITLE
[KWS]Add kws example on HeySnips dataset.

### PR DESCRIPTION
**Pull request by KPatr1ck**
### PR types

New features
### PR changes

Models
### Describe

Add mdtc model for kws.

### CODE REVIEW

**Review by zh794390558 — COMMENTED** [review]
对应的paper是？

File: examples/hey_snips/kws0/conf/mdtc.yaml [hunk 1]
```diff
@@ -0,0 +1,39 @@
+data:
```
Comment (zh794390558) [inline examples/hey_snips/kws0/conf/mdtc.yaml:1]: 配置文件参考其他egs，展开吧。不建议用层级的了。

File: examples/hey_snips/kws0/local/score.sh [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+
```
Comment (zh794390558) [inline examples/hey_snips/kws0/local/score.sh:2]: 加上usage，并check输入

File: paddleaudio/paddleaudio/datasets/dataset.py [hunk 3]
```diff
@@ -73,16 +77,24 @@ def _convert_to_record(self, idx):
         feat_func = feat_funcs[self.feat_type]
 
         record = {}
-        record['feat'] = feat_func(
-            waveform, sample_rate,
-            **self.feat_config) if feat_func else waveform
+        if self.feat_type in ['kaldi_fbank', 'kaldi_mfcc']:
+            waveform = paddle.to_tensor(waveform).unsqueeze(0)  # (C, T)
+            record['feat'] = feat_func(
+                waveform=waveform, sr=self.sample_rate, **self.feat_config)
```
Comment (zh794390558) [inline paddleaudio/paddleaudio/datasets/dataset.py]: 需要的参数需要对应指出，不建议使用**

File: paddlespeech/kws/exps/mdtc/compute_det.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+import os
+
+import yaml
+from tqdm import tqdm
+
+from paddlespeech.s2t.utils.dynamic_import import dynamic_import
+
+# yapf: disable
+parser = argparse.ArgumentParser(__doc__)
+parser.add_argument("--cfg_path", type=str, required=True)
```
Comment (zh794390558) [inline paddlespeech/kws/exps/mdtc/compute_det.py:26]: 不建议用配置文件传入参数，需要单独指出

File: paddlespeech/kws/exps/mdtc/compute_det.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+parser.add_argument('--keyword', type=int, default=0, help='keyword label')
+parser.add_argument('--step', type=float, default=0.01, help='threshold step')
+parser.add_argument('--window_shift', type=int, default=50, help='window_shift is used to skip the frames after triggered')
+args = parser.parse_args()
+# yapf: enable
+
+
+def load_label_and_score(keyword, ds, score_file):
```
Comment (zh794390558) [inline paddlespeech/kws/exps/mdtc/compute_det.py]: ds是什么？加上typehint

File: paddlespeech/kws/exps/mdtc/compute_det.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+import yaml
+from tqdm import tqdm
+
+from paddlespeech.s2t.utils.dynamic_import import dynamic_import
+
+# yapf: disable
+parser = argparse.ArgumentParser(__doc__)
+parser.add_argument("--cfg_path", type=str, required=True)
+parser.add_argument('--keyword', type=int, default=0, help='keyword label')
```
Comment (zh794390558) [inline paddlespeech/kws/exps/mdtc/compute_det.py]: keyword_id

File: paddlespeech/kws/exps/mdtc/compute_det.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+parser.add_argument('--step', type=float, default=0.01, help='threshold step')
+parser.add_argument('--window_shift', type=int, default=50, help='window_shift is used to skip the frames after triggered')
```
Comment (zh794390558) [inline paddlespeech/kws/exps/mdtc/compute_det.py:29]: 名字看不出含义

File: paddlespeech/kws/exps/mdtc/compute_det.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        for line in fin:
+            arr = line.strip().split()
+            key = arr[0]
+            current_keyword = arr[1]
+            str_list = arr[2:]
+            if int(current_keyword) == keyword:
+                scores = list(map(float, str_list))
+                if key not in score_table:
+                    score_table.update({key: scores})
+    keyword_table = {}
```
Comment (zh794390558) [inline paddlespeech/kws/exps/mdtc/compute_det.py]: 描述下这两个存放的是什么格式数据

File: paddlespeech/kws/exps/mdtc/compute_det.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+                if float(score) < threshold:
+                    num_false_reject += 1
+            num_false_alarm = 0
+            # transverse the all filler_table
+            for key, score_list in filler_table.items():
+                i = 0
+                while i < len(score_list):
+                    if score_list[i] >= threshold:
+                        num_false_alarm += 1
+                        i += args.window_shift
```
Comment (zh794390558) [inline paddlespeech/kws/exps/mdtc/compute_det.py:101]: 为什么要skip？

File: paddlespeech/kws/exps/mdtc/plot_det_curve.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+import argparse
+import os
+
+import matplotlib.pyplot as plt
+import numpy as np
+import yaml
+
+# yapf: disable
+parser = argparse.ArgumentParser(__doc__)
+parser.add_argument("--cfg_path", type=str, required=True)
```
Comment (zh794390558) [inline paddlespeech/kws/exps/mdtc/plot_det_curve.py:24]: 同样不用config

File: paddlespeech/kws/exps/mdtc/score.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+import yaml
+from tqdm import tqdm
+
+from paddlespeech.kws.exps.mdtc.collate import collate_features
+from paddlespeech.kws.models.mdtc import KWSModel
+from paddlespeech.s2t.utils.dynamic_import import dynamic_import
+
+# yapf: disable
+parser = argparse.ArgumentParser(__doc__)
+parser.add_argument("--cfg_path", type=str, required=True)
```
Comment (zh794390558) [inline paddlespeech/kws/exps/mdtc/score.py:28]: 同上

File: paddlespeech/kws/exps/mdtc/score.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+import paddle
+import yaml
+from tqdm import tqdm
+
+from paddlespeech.kws.exps.mdtc.collate import collate_features
+from paddlespeech.kws.models.mdtc import KWSModel
+from paddlespeech.s2t.utils.dynamic_import import dynamic_import
+
+# yapf: disable
+parser = argparse.ArgumentParser(__doc__)
```
Comment (zh794390558) [inline paddlespeech/kws/exps/mdtc/score.py:27]: 用文字描述功能

File: paddlespeech/kws/exps/mdtc/train.py [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+                batch_sampler=dev_sampler,
+                num_workers=training_conf['num_workers'],
+                return_list=True,
+                use_buffer_reader=True,
+                collate_fn=collate_features, )
+
+            model.eval()
+            num_corrects = 0
+            num_samples = 0
+            with logger.processing('Evaluation on validation dataset'):
```
Comment (zh794390558) [inline paddlespeech/kws/exps/mdtc/train.py:147]: 这是什么操作？

File: paddlespeech/kws/exps/mdtc/train.py [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+                    logits = model(feats)
+                    loss, corrects, acc = criterion(logits, labels, lengths)
+                    num_corrects += corrects
+                    num_samples += feats.shape[0]
+
+            eval_acc = num_corrects / num_samples
+            print_msg = '[Evaluation result]'
+            print_msg += ' dev_acc={:.4f}'.format(eval_acc)
+
+            logger.eval(print_msg)
```
Comment (zh794390558) [inline paddlespeech/kws/exps/mdtc/train.py:159]: 还有这个？

File: paddlespeech/kws/models/loss.py [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+            num_correct += 1
+        # Predict correct as the filler, filler id < 0
+        if max_p < 0.5 and target[i].item() < 0:
+            num_correct += 1
+    acc = num_correct / num_utts
+    # acc = 0.0
+    return loss, num_correct, acc
+
+
+def padding_mask(lengths: paddle.Tensor) -> paddle.Tensor:
```
Comment (zh794390558) [inline paddlespeech/kws/models/loss.py]: 放到开头

File: paddlespeech/kws/models/loss.py [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+
+
+def max_pooling_loss(logits: paddle.Tensor,
+                     target: paddle.Tensor,
+                     lengths: paddle.Tensor,
+                     min_duration: int=0):
+
+    mask = padding_mask(lengths)
+    num_utts = logits.shape[0]
+    num_keywords = logits.shape[2]
```
Comment (zh794390558) [inline paddlespeech/kws/models/loss.py:40]: num_keywords 好奇怪。 
B，T，D

File: paddlespeech/kws/models/loss.py [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+
+    loss = 0.0
+    for i in range(num_utts):
```
Comment (zh794390558) [inline paddlespeech/kws/models/loss.py:43]: 为什么这个不直接调用CE？

**Comment by KPatr1ck** [comment]
> 对应的paper是？

在README中的模型处给出了，https://arxiv.org/pdf/2102.13552.pdf

File: paddlespeech/kws/models/loss.py (hunk 17 above)
Comment (KPatr1ck) [reply paddlespeech/kws/models/loss.py:43]: 不可以直接用CrossEntropy，需要在有效的音频长度内取max作为得分再算CE。

File: paddlespeech/kws/models/loss.py (hunk 15 above)
Comment (KPatr1ck) [reply paddlespeech/kws/models/loss.py]: done

File: paddlespeech/kws/exps/mdtc/train.py (hunk 13 above)
Comment (KPatr1ck) [reply paddlespeech/kws/exps/mdtc/train.py:147]: logger显示优化

File: paddlespeech/kws/exps/mdtc/train.py (hunk 14 above)
Comment (KPatr1ck) [reply paddlespeech/kws/exps/mdtc/train.py:159]: 同上。paddleaudio.utils.logger

File: paddleaudio/paddleaudio/datasets/dataset.py (hunk 3 above)
Comment (KPatr1ck) [reply paddleaudio/paddleaudio/datasets/dataset.py]: feature api的参数太多，且大部分为默认，我加个说明让用户根据api中的kwargs去配置。

File: paddlespeech/kws/exps/mdtc/compute_det.py (hunk 5 above)
Comment (KPatr1ck) [reply paddlespeech/kws/exps/mdtc/compute_det.py]: done

File: paddlespeech/kws/exps/mdtc/compute_det.py (hunk 6 above)
Comment (KPatr1ck) [reply paddlespeech/kws/exps/mdtc/compute_det.py]: done

File: paddlespeech/kws/exps/mdtc/compute_det.py (hunk 8 above)
Comment (KPatr1ck) [reply paddlespeech/kws/exps/mdtc/compute_det.py]: done

File: paddlespeech/kws/exps/mdtc/compute_det.py (hunk 7 above)
Comment (KPatr1ck) [reply paddlespeech/kws/exps/mdtc/compute_det.py:29]: done

File: paddlespeech/kws/exps/mdtc/compute_det.py (hunk 9 above)
Comment (KPatr1ck) [reply paddlespeech/kws/exps/mdtc/compute_det.py:101]: 可以理解为短时间内不可能有2次唤醒，50*25ms=0.5秒，这个期间为冷却期。

File: examples/hey_snips/kws0/local/score.sh (hunk 2 above)
Comment (KPatr1ck) [reply examples/hey_snips/kws0/local/score.sh:2]: 这个脚本不推荐用户单独使用，正常都是从run.sh调用，我在run里加检查

File: paddlespeech/kws/exps/mdtc/compute_det.py (hunk 9 above)
Comment (zh794390558) [reply paddlespeech/kws/exps/mdtc/compute_det.py:101]: 50*10=500ms

File: paddlespeech/kws/models/loss.py [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+        for j in range(num_keywords):
+            # Add entropy loss CE = -(t * log(p) + (1 - t) * log(1 - p))
+            if target[i] == j:
+                # For the keyword, do max-polling
```
Comment (zh794390558) [inline paddlespeech/kws/models/loss.py:47]: 看着是在T上做了max_pool后算Cross Entropy

File: paddlespeech/kws/models/mdtc.py [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+        outputs = outputs.transpose([0, 2, 1])
+        return outputs, None
+
+
+class KWSModel(nn.Layer):
+    def __init__(self, backbone, num_keywords):
+        super(KWSModel, self).__init__()
+        self.backbone = backbone
+        self.linear = nn.Linear(self.backbone.hidden_dim, num_keywords)
+        self.activation = nn.Sigmoid()
```
Comment (zh794390558) [inline paddlespeech/kws/models/mdtc.py:228]: 二分类

File: paddlespeech/kws/models/mdtc.py [hunk 20]
```diff
@@ -0,0 +1,233 @@
...
+            kernel_size: int,
+            dilation: int,
+            causal: bool, ):
+        super(TCNBlock, self).__init__()
+        self.in_channels = in_channels
+        self.res_channels = res_channels
+        self.kernel_size = kernel_size
+        self.dilation = dilation
+        self.causal = causal
+        self.receptive_fields = dilation * (kernel_size - 1)
```
Comment (zh794390558) [inline paddlespeech/kws/models/mdtc.py:70]: kernel_size - 1 if no_dilation else dilation * (kernel_size -1)

**Review by zh794390558 — APPROVED** [review]
LGTM
